### PR TITLE
feat(x/ecocredit): add basket state migration

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Ref: #1358 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by technicallyty — APPROVED** [review]
LGTM! nice to be able to use the native register migrations 👍🏻

**Review by ryanchristo — COMMENTED** [review]
Tested an upgrade from v4 to v5 and it looks like the upgrade resets the basket fees param to an empty array. Tested updating the param with a chain that start on v5 and it works as expected.

https://hackmd.io/MP6BaYA_SOiQtEomPD8ong

Also, I believe we should be using the consensus version rather than the regen-ledger version for the migration name, same as the sdk. The sdk is still using the sdk version in some modules but are migrating to the use of the consensus version.

https://github.com/cosmos/cosmos-sdk/tree/main/x/gov/migrations

**Comment by ryanchristo** [comment]
For reference (use of consensus version and using migrator):

https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-041-in-place-store-migrations.md

**Comment by aleem1314** [comment]
> Tested an upgrade from v4 to v5 and it looks like the upgrade resets the basket fees param to an empty array. Tested updating the param with a chain that start on v5 and it works as expected.

Nice catch. The issue is when we start a new chain SDK module manager sets a consensus version for registered modules (we are not using SDK's module manager for regen modules). The ecocredit module migrations not running because there is no consensus version present in the state. I've updated the upgrade handler to initialize the consensus version for `ecocredit` and `data` modules.

File: x/ecocredit/migrations/v3/state.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	basev1beta1 "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+
+	basketapi "github.com/regen-network/regen-ledger/api/regen/ecocredit/basket/v1"
+	api "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+// MigrateState performs in-place store migrations from v4.0 to v5.0.
```
Comment (ryanchristo) [inline x/ecocredit/migrations/v3/state.go:13]:
```suggestion
// MigrateState performs in-place store migrations from ConsensusVersion 2 to 3.
```

**Review by ryanchristo — APPROVED** [review]
[tACK](https://hackmd.io/MP6BaYA_SOiQtEomPD8ong)

**Pull request merged**